vertical-pod-autoscaler/enhancements/4566-min-replicas/README.md

```diff
@@ -0,0 +1,107 @@
+- [Implementation History](#implementation-history)
+- [Alternatives](#alternatives)
+   - [Existing Behaviour](#existing-behaviour)
+   - [Reuse Cluster Autoscaler Annotations](#reuse-cluster-autoscaler-annotations)
+<!-- /toc -->
+
+## Summary
+
+The default behaviour of VPA Updater is to allow Pod eviction only if there are
+at least 2 live replicas, in order to avoid temporaty total unavailability of a
+workload under VPA in Auto mode. This can be changed globally with
+`--min-replicas` flag. However such a change might be deemed dangerous in a
+centrally-managed multi-tenant cluster.
+
+This proposal addresses the problem by allowing to specify per VPA object a
+minimum number of replicas required to start Pod eviction (thus active VPA
+actuation). This allows to keep the current relatively safe default and override
+it explicitly only where needed.
+
+## Motivation
+
+The motivation behind the change is due to VPA users needing the configuration
+option in managed environments to allow for eviction of a single replica, see
+[issue #3986](https://github.com/kubernetes/autoscaler/issues/3986) or
+[issue #1828](https://github.com/kubernetes/autoscaler/issues/1828).
+
+### Goals
+
+- Main: allow workload owner to use VPA in Auto/Recreate mode on a single-replica
+  workloads,
+- Secondary: allow workload owner to specify higher number of replicas required
+  to be kept alive. Note that this can also be achieved with
+  [PDB](https://kubernetes.io/docs/tasks/run-application/configure-pdb/).
+
+### Non-Goals
+
+- Create any advanced form of policy around VPA-introduced disruption.
+
+## Proposal
+
+The proposal is to add `minReplicas` field under VPA `spec.updatePolicy` and
+alter VPA Updater behaviour so it respects the new field:
+- if the field is set, use its value instead of the global `--min-replicas`
+  flag, only for this particular VPA object,
+- if the field is not set, use the global `--min-replicas` flag, keeping
+  backward compatibility.
+
+The field is ignored by other VPA components (Recommender, Admission
+ontroller).
+
+Since the change is backward-compatible the suggestion is to simply extend `v1`
+version of VPA API, avoiding the hassle of introducing a new API version.
+
+## Design Details
+
+Suggested implementation is present in [PR
+4560](https://github.com/kubernetes/autoscaler/pull/4560).
+
+### Test Plan
+
+Add automated tests that update VPA objects, altering the value of
+`spec.updatePolicy.minReplicas` flag and verifying that the behaviour of VPA
+Updater changes accordingly.
+
+## Implementation History
+
+- 2021-12-28: initial version
+
+## Alternatives
+
+### Existing Behaviour
+
+The only existing alternative to achieve the behaviour allowed by the proposed
+change is to set global `--min-replicas` flag to `1` and Pod Disruption Budget
+to protect single-replica workloads from being disrupted, if needed.
+
+While this would end up in a very similar (if not identical) cluster behaviour,
```

Comment:
I think cluster behavior is different.

With this KEP we can protect 1 pod controllers from eviction by default (and can override that per-VPA).
With the alternative described in this section we evict 1-pod controllers by default (and can override that per-controller).

I'd rephrase to something like:

While this allows the same configuration for each individual controller it changes the default behavior to an unsafe one. As a result it requires cluster-wide coordination...